File: JPP/lab-1/z6/src/iter.c
#include "lib.h"

// PRIVATE

void swap(int* a, int* b)
{
    const int temp = *a;
    *a = *b;
    *b = temp;
}

Pair egcd(int a, int b)
{
    Pair result = {1,0};
    Pair prev = {0,1};

    while (b != 0) {
        const int q = a / b;

        const int x = result.x - q * prev.x;
        result.x = prev.x;
        prev.x = x;

        const int y = result.y - q * prev.y;
        result.y = prev.y;
        prev.y = y;

        const int temp = a;
        a = b;
        b = temp - q * b;
    }

    return result;
}

// PUBLIC

int factorial(const int n)
{
    int val = 1;
    
    for (int i = 1; i <= n; i++)
        val *= i;

    return val;
}

int gcd(int a, int b)
{
    if (a < b)
        swap(&a, &b);

    while (b != 0)
    {
        int temp = a;
        a = b;
        b = temp % b;
    }

    return a;
}

Pair solve(int a, int b, int c)
{
    Pair result = {0,0};

    const int d = gcd(a,b);

    if (c % d == 0)
    {
        const Pair st = egcd(a,b);

        const int s = st.x;
        const int t = st.y;
        const int k = c / d;

        result.x = s*k;
        result.y = t*k;
    }

    return result;
}

File: JPP/lab-1/z7/src/iter.rs
pub mod lib {
    pub struct Pair {
        pub x: i32,
        pub y: i32,
    }

    pub fn factorial(n: i32) -> i32 {
        let mut result = 1;
        for i in 1 .. n+1 {
            result *= i;
        }

        return result;
    }

    pub fn gcd(mut a: i32, mut b: i32) -> i32 {

        if a < b {
            std::mem::swap(&mut a, &mut b);
        }

        while b != 0 {
            a = a % b;
            std::mem::swap(&mut a, &mut b);
        }

        return a;
    }

    pub fn solve(a: i32, b: i32, c: i32) -> Pair {
        let mut result = Pair {x: 0, y: 0};

        let d = gcd(a, b);

        if c % d == 0 {
            result = egcd(a, b);
            result.x *= c / d;
            result.y *= c / d;
        }

        return result;
    }

    fn egcd(mut a: i32, mut b: i32)  -> Pair {
        let mut result = Pair {x: 1, y: 0};
        let mut prev = Pair {x: 0, y: 1};

        while b != 0 {
            let q = a / b;

            let temp = Pair {x: result.x, y: result.y};
            result.x = prev.x;
            result.y = prev.y;

            prev.x = temp.x - q * prev.x;
            prev.y = temp.y - q * prev.y;

            let temp = a;
            a = b;
            b = temp % b;
        }

        return result;
    }
}